stack/corepersistence/queryindex/src/main/java/org/apache/usergrid/persistence/index/impl/EsIndexProducerImpl.java

```diff
@@ -0,0 +1,209 @@
+
+        //wire up the gauge of inflight messages
+        metricsFactory.addGauge(EsIndexProducerImpl.class, "index_buffer.inflight", () -> inFlight.longValue());
+
+
+        this.indexTimer = metricsFactory.getTimer( EsIndexProducerImpl.class, "index" );
+
+        this.config = config;
+        this.failureMonitor = new FailureMonitorImpl(config, provider);
+        this.client = provider.getClient();
+        this.indexFig = indexFig;
+
+
+        //batch up sets of some size and send them in batch
+
+    }
+
+    public Observable<IndexOperationMessage>  put( IndexOperationMessage message ) {
+        Preconditions.checkNotNull(message, "Message cannot be null");
+        indexSizeCounter.inc(message.getDeIndexRequests().size());
+        indexSizeCounter.inc(message.getIndexRequests().size());
+        return  processBatch(message);
+    }
+
+
+    /**
+     * Process the buffer of batches
+     * @param batch
+     * @return
+     */
+    private Observable<IndexOperationMessage> processBatch( final IndexOperationMessage batch ) {
+
+        //take our stream of batches, then stream then into individual ops for consumption on ES
+        final Set<IndexOperation> indexOperationSet = batch.getIndexRequests();
+        final Set<DeIndexOperation> deIndexOperationSet = batch.getDeIndexRequests();
+
+        final int indexOperationSetSize = indexOperationSet.size();
+        final int deIndexOperationSetSize = deIndexOperationSet.size();
+
+        log.debug("Emitting {} add and {} remove operations", indexOperationSetSize, deIndexOperationSetSize);
+
+        indexSizeCounter.dec(indexOperationSetSize);
+        indexSizeCounter.dec(deIndexOperationSetSize);
+
+        final Observable<IndexOperation> index = Observable.from(batch.getIndexRequests());
+        final Observable<DeIndexOperation> deIndex = Observable.from(batch.getDeIndexRequests());
+
+        final Observable<BatchOperation> batchOps = Observable.merge(index, deIndex);
+
+        //buffer into the max size we can send ES and fire them all off until we're completed
+        final Observable<BulkRequestBuilder> requests = batchOps.buffer(indexFig.getIndexBatchSize())
+            //flatten the buffer into a single batch execution
+            .flatMap(individualOps -> Observable.from(individualOps)
+                //collect them
+                .collect(() -> initRequest(), (bulkRequestBuilder, batchOperation) -> {
+                    log.debug("adding operation {} to bulkRequestBuilder {}", batchOperation, bulkRequestBuilder);
+                    batchOperation.doOperation(client, bulkRequestBuilder);
+                }))
+                //write them
+            .doOnNext(bulkRequestBuilder -> sendRequest(bulkRequestBuilder));
+
+
+        //now that we've processed them all, ack the futures after our last batch comes through
+        final Observable<IndexOperationMessage> processedIndexOperations =
+            requests.lastOrDefault(null).flatMap(lastRequest -> {
```

Comment:
I think you can get rid of this lastOrDefaultRight?  Won't it always contain the batch?